src/Go/boardAnalysis/goAI.ts

```diff
@@ -15,20 +15,29 @@ import {
   getAllEyesByChainId,
   getAllNeighboringChains,
   getAllValidMoves,
+  getPreviousMoveDetails,
 } from "./boardAnalysis";
 import { findDisputedTerritory } from "./controlledTerritory";
 import { findAnyMatchedPatterns } from "./patternMatching";
 import { WHRNG } from "../../Casino/RNG";
 import { Go, GoEvents } from "../Go";
 
 let currentAITurn: Promise<Play> | null = null;
+let currentTurnResolver: ((value: Play | PromiseLike<Play>) => void) | null = null;
 
 /**
  * Retrieves a move from the current faction in response to the player's move
  */
 export function makeAIMove(boardState: BoardState): Promise<Play> {
   // If AI is already taking their turn, return the existing turn.
-  if (currentAITurn) return currentAITurn;
+  if (currentAITurn) {
+    return Go.nextTurn;
+  }
+  if (boardState.ai === GoOpponent.none) {
+    GoEvents.emit();
+    return (currentAITurn = Go.nextTurn = new Promise((resolve) => (currentTurnResolver = resolve)));
```

Comment:
This bypasses the finally, so it never resets currentAITurn. You're also emitting GoEvents synchronously, instead of when the currentTurnResolver resolves?

I think what you want is to conditionally construct either this resolved promise, *or* the `getMove.then` promise below. And then either way, return `that.finally(...)`.

Reply:
The promise needs to not resolve here, though, nor set currentAITurn, yet. It waits until resolveCurrentTurn() is called externally, rather than following the code below. In this case the AI we are waiting on is the player, manually playing white against their script (e.g. a human brain made by humans, which is technically an intelligence and artificial XD).

Reply:
I misspoke (again), but I would have thought the Idea would be clear from the code. The promise you are constructing here is *not* resolved (yet), since it is triggered by `currentTurnResolver`.

Later on, you do some custom stuff to perform the action of "when `currentTurnResolver` is called, reset `currentAITurn`." But this is duplicate code; this is already handled by the `finally` promise that chains off the main one. The main promise is not resolved yet either; both are the same in that they are waiting for the turn to be done. This is why it makes sense for them to share the finally.

Reply:
I've changed this to apply the finally() to Go.nextTurn regardless of which branch of the Ai/No AI logic is followed. Let me know if this is closer to what you were thinking.